k2/csrc/properties.cc

```diff
@@ -78,6 +79,51 @@ bool HasSelfLoops(const Fsa &fsa) {
   return false;
 }
 
+bool CheckCycles(StateId s, std::vector<bool> *visited,
+                 std::vector<bool> *back_arc, const std::vector<Arc> &arcs) {
+  if ((*visited)[s] == false) {
```

Comment:
Seems the implentation  will iterate over the arcs multiple times which is really inefficient. 
BTW, I strongly suggest we should try to avoid recursion

Reply:
A directed graph is acyclic iff it has no strongly connected subgraph with more than one state. Since we already have an implementation of Tarjan's scc algorithm, that should be enough for detecting IsAcyclic.
In fact, it seems that the [ConnectCore function](https://github.com/danpovey/k2/blob/152ced3082914ffd1dfaeba71d2e5e11949678a2/k2/csrc/fsa_algo.cc#L62) does the job of detecting IsAcyclic; it returns true if the fsa is acyclic and false if it is cyclic. Do we still need to implement a separate function IsAcyclic?

Reply:
function `CheckCycles` is not needed anymore, right? I guess you can remove it.